vortex-array/src/arrays/primitive/compute/take/avx2.rs

```diff
@@ -0,0 +1,350 @@
+                match_each_integer_ptype!(indices.ptype(), |I| {
+                    // NOTE: we don't need to pre-scan the indices here, because
+                    //  the slice indexing inside of take_primitive_scalar will do bounds checks.
+                    let result =
+                        take_primitive_scalar(values.as_slice::<T>(), indices.as_slice::<I>());
+                    Ok(PrimitiveArray::new(result, validity).into_array())
+                })
+            })
+        }
+    }
+}
+
+/// The main gather function that is used by the inner loop kernel for AVX2 gather.
+pub(crate) trait GatherFn<Idx, Values> {
+    /// The number of data elements that are written to the `dst` on each loop iteration.
+    const WIDTH: usize;
+    /// The number of indices read from `indices` on each loop iteration.
+    /// Depending on the available instructions and bit-width we may stride by a larger amount
+    /// than we actually end up reading from `src` (governed by the `WIDTH` parameter).
+    const STRIDE: usize = Self::WIDTH;
+
+    /// Gather values from `src` into the `dst` using the `indices`, optionally using
+    /// SIMD instructions.
+    ///
+    /// # Safety
+    ///
+    /// This function can read up to `STRIDE` elements through `indices`, and read/write up to
+    /// `WIDTH` elements through `src` and `dst` respectively.
+    unsafe fn gather(indices: *const Idx, src: *const Values, dst: *mut Values);
+}
+
+/// AVX2 version of GatherFn defined for 32- and 64-bit value types.
+enum AVX2Gather {}
+
+unsafe fn identity<T>(input: T) -> T {
+    input
+}
+
+macro_rules! impl_gather {
+    ($idx:ty, $({$value:ty => load: $load:ident, extend: $extend:ident, gather: $gather:ident, store: $store:ident, WIDTH = $WIDTH:literal, STRIDE = $STRIDE:literal }),+) => {
+        $(
+            impl GatherFn<$idx, $value> for AVX2Gather {
+                const WIDTH: usize = $WIDTH;
+                const STRIDE: usize = $STRIDE;
+
+                #[allow(clippy::cast_possible_truncation)]
+                #[inline(always)]
+                unsafe fn gather(indices: *const $idx, src: *const $value, dst: *mut $value) {
+                    const {
+                        assert!($WIDTH <= $STRIDE, "dst cannot advance by more than the stride");
+                    }
+
+                    const SCALE: i32 = std::mem::size_of::<$value>() as i32;
+
+                    let indices_vec = unsafe { $load(indices.cast()) };
+                    // Extend indices to fill vector register
+                    let indices_vec = unsafe { $extend(indices_vec) };
+                    // Gather the values into vector register
+                    let values_vec = unsafe { $gather::<SCALE>(src.cast(), indices_vec) };
+
+                    // Write the vec out to dst.
+                    unsafe { $store(dst.cast(), values_vec) };
+                }
+            }
+        )*
+    };
+}
+
+impl_gather!(u8,
```

Comment:
we should implement f32 and f64 as value types since there are relevant gather instructions for them. This PR is probably enough to review so I'd prefer to do in follow up